File: packages/db/src/crypto.ts
// Credential crypto for @hybrid/db. AES-256-GCM via node:crypto.
//
// Guards the `credentials` jsonb columns on payment_account / courier_account.
// Plaintext credential maps (e.g. {app_key, app_secret} for a gateway) are
// sealed into a self-describing JSON envelope (SealedSecret) before they ever
// touch the DB, and opened only server-side inside a withTenant() transaction.
//
// Pure util: no DB, no Next, no logging of secrets. Callers run inside withTenant.
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

// JSON envelope stored in jsonb. Versioned so the algorithm can evolve without
// ambiguity. All binary fields are base64. `v:1` + `alg:"A256GCM"` is the only
// shape this module emits or accepts today.
export interface SealedSecret {
  v: 1;
  alg: "A256GCM";
  iv: string; // base64, 12 bytes (GCM standard nonce length)
  ct: string; // base64 ciphertext
  tag: string; // base64, 16-byte GCM auth tag
}

const ALG = "aes-256-gcm";
const KEY_BYTES = 32; // AES-256
const IV_BYTES = 12; // GCM standard nonce
const TAG_BYTES = 16; // GCM auth tag

// Fail-fast key load, mirroring the devSessionSecret() pattern: resolved on
// first use (not at import) and validated to be exactly a base64-encoded
// 32-byte key. A missing or wrong-length key throws immediately rather than
// silently encrypting with a guessable/short key.
let cachedKey: Buffer | null = null;

export function encryptionKey(): Buffer {
  if (cachedKey) return cachedKey;

  const raw = process.env.APP_ENCRYPTION_KEY;
  if (!raw) {
    throw new Error("APP_ENCRYPTION_KEY is not set (required for credential crypto)");
  }

  let key: Buffer;
  try {
    key = Buffer.from(raw, "base64");
  } catch {
    throw new Error("APP_ENCRYPTION_KEY must be base64-encoded");
  }

  if (key.length !== KEY_BYTES) {
    throw new Error(
      `APP_ENCRYPTION_KEY must decode to ${KEY_BYTES} bytes (got ${key.length}); ` +
        "generate one with: openssl rand -base64 32",
    );
  }

  cachedKey = key;
  return key;
}

// Seal a plaintext credential map into a SealedSecret envelope. A fresh random
// 12-byte IV is generated per call, so identical inputs never produce identical
// ciphertext.
export function sealCredentials(plain: Record<string, string>): SealedSecret {
  const key = encryptionKey();
  const iv = randomBytes(IV_BYTES);
  const cipher = createCipheriv(ALG, key, iv);

  const plaintext = Buffer.from(JSON.stringify(plain), "utf8");
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();

  return {
    v: 1,
    alg: "A256GCM",
    iv: iv.toString("base64"),
    ct: ct.toString("base64"),
    tag: tag.toString("base64"),
  };
}

// Open a SealedSecret back to its plaintext map. GCM authentication means a
// tampered ciphertext or tag, or the wrong key, causes decipher.final() to
// throw — the result is never silently corrupt.
export function openCredentials(sealed: SealedSecret): Record<string, string> {
  if (!isSealed(sealed)) {
    throw new Error("openCredentials: value is not a SealedSecret envelope");
  }

  const key = encryptionKey();
  const iv = Buffer.from(sealed.iv, "base64");
  const tag = Buffer.from(sealed.tag, "base64");

  if (iv.length !== IV_BYTES) {
    throw new Error("SealedSecret.iv has invalid length");
  }
  if (tag.length !== TAG_BYTES) {
    throw new Error("SealedSecret.tag has invalid length");
  }

  const decipher = createDecipheriv(ALG, key, iv);
  decipher.setAuthTag(tag);

  const ct = Buffer.from(sealed.ct, "base64");
  const plaintext = Buffer.concat([decipher.update(ct), decipher.final()]);

  return JSON.parse(plaintext.toString("utf8")) as Record<string, string>;
}

// Type guard: is this unknown value (e.g. a jsonb column read) a SealedSecret?
export function isSealed(v: unknown): v is SealedSecret {
  if (typeof v !== "object" || v === null) return false;
  const o = v as Record<string, unknown>;
  return (
    o.v === 1 &&
    o.alg === "A256GCM" &&
    typeof o.iv === "string" &&
    typeof o.ct === "string" &&
    typeof o.tag === "string"
  );
}

File: apps/api/app/crypto.py
"""Credential crypto — AES-256-GCM, byte-compatible with packages/db/src/crypto.ts.

The web app seals gateway/courier credential maps into a versioned JSON envelope
(`SealedSecret`) before they touch the DB; this opens them server-side for a job.

Envelope (jsonb): {v:1, alg:"A256GCM", iv:b64(12), ct:b64, tag:b64(16)}.
Node's GCM exposes ciphertext and the 16-byte auth tag separately; Python's
AESGCM expects them concatenated (ct || tag) — that is the only translation.
"""
import base64
import json
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .config import get_settings

KEY_BYTES = 32
IV_BYTES = 12
TAG_BYTES = 16


def _key() -> bytes:
    raw = get_settings().app_encryption_key
    try:
        key = base64.b64decode(raw, validate=True)
    except Exception as exc:  # noqa: BLE001 - re-raised with a clear message
        raise ValueError("APP_ENCRYPTION_KEY must be base64-encoded") from exc
    if len(key) != KEY_BYTES:
        raise ValueError(
            f"APP_ENCRYPTION_KEY must decode to {KEY_BYTES} bytes (got {len(key)}); "
            "generate with: openssl rand -base64 32"
        )
    return key


def is_sealed(value: Any) -> bool:
    """Type guard: does this (e.g. a jsonb column read) look like a SealedSecret?"""
    if not isinstance(value, dict):
        return False
    return (
        value.get("v") == 1
        and value.get("alg") == "A256GCM"
        and all(isinstance(value.get(k), str) for k in ("iv", "ct", "tag"))
    )


def open_credentials(sealed: dict[str, Any]) -> dict[str, str]:
    """Open a SealedSecret back to its plaintext map. A tampered ct/tag or the
    wrong key raises (GCM authentication) — never returns silently-corrupt data."""
    if not is_sealed(sealed):
        raise ValueError("open_credentials: value is not a SealedSecret envelope")

    iv = base64.b64decode(sealed["iv"])
    tag = base64.b64decode(sealed["tag"])
    ct = base64.b64decode(sealed["ct"])
    if len(iv) != IV_BYTES:
        raise ValueError("SealedSecret.iv has invalid length")
    if len(tag) != TAG_BYTES:
        raise ValueError("SealedSecret.tag has invalid length")

    plaintext = AESGCM(_key()).decrypt(iv, ct + tag, None)
    return json.loads(plaintext.decode("utf-8"))


def seal_credentials(plain: dict[str, str]) -> dict[str, Any]:
    """Seal a plaintext credential map (fresh 12-byte IV per call). Emitted
    envelope is openable by openCredentials() in crypto.ts."""
    iv = os.urandom(IV_BYTES)
    # Compact separators so the JSON matches JSON.stringify's default spacing.
    blob = AESGCM(_key()).encrypt(
        iv, json.dumps(plain, separators=(",", ":")).encode("utf-8"), None
    )
    ct, tag = blob[:-TAG_BYTES], blob[-TAG_BYTES:]
    return {
        "v": 1,
        "alg": "A256GCM",
        "iv": base64.b64encode(iv).decode("ascii"),
        "ct": base64.b64encode(ct).decode("ascii"),
        "tag": base64.b64encode(tag).decode("ascii"),
    }
